be/src/exec/vectorized/partition/partition_hash_map.h

```diff
@@ -0,0 +1,250 @@
+
+    void append_chunk(ChunkPtr chunk, const Columns& key_columns) {
+        DCHECK(!key_columns[0]->is_nullable());
+        const auto* key_column = down_cast<ColumnType*>(key_columns[0].get());
+        const auto& key_column_data = key_column->get_data();
+        append_chunk_for_one_key(hash_map, chunk, [&](uint32_t offset) { return key_column_data[offset]; });
+    }
+};
+
+template <PrimitiveType primitive_type, typename HashMap>
+struct PartitionHashMapWithOneNullableNumberKey : public PartitionHashMapBase {
+    using Iterator = typename HashMap::iterator;
+    using ColumnType = RunTimeColumnType<primitive_type>;
+    using FieldType = RunTimeCppType<primitive_type>;
+    HashMap hash_map;
+    PartitionChunks null_key_value;
+
+    PartitionHashMapWithOneNullableNumberKey(int32_t chunk_size) : PartitionHashMapBase(chunk_size) {}
+
+    void append_chunk(ChunkPtr chunk, const Columns& key_columns) {
+        DCHECK(key_columns[0]->is_nullable());
+        const auto* nullable_key_column = ColumnHelper::as_raw_column<NullableColumn>(key_columns[0].get());
+        const auto& key_column_data = down_cast<ColumnType*>(nullable_key_column->data_column().get())->get_data();
+        append_chunk_for_one_nullable_key(hash_map, null_key_value, chunk, nullable_key_column,
+                                          [&](uint32_t offset) { return key_column_data[offset]; });
+    }
+};
+
+template <typename HashMap>
+struct PartitionHashMapWithOneStringKey : public PartitionHashMapBase {
+    using Iterator = typename HashMap::iterator;
+    HashMap hash_map;
+
+    PartitionHashMapWithOneStringKey(int32_t chunk_size) : PartitionHashMapBase(chunk_size) {}
+
+    void append_chunk(ChunkPtr chunk, const Columns& key_columns) {
+        DCHECK(!key_columns[0]->is_nullable());
+        const auto* key_column = down_cast<BinaryColumn*>(key_columns[0].get());
+        append_chunk_for_one_key(hash_map, chunk, [&](uint32_t offset) { return key_column->get_slice(offset); });
+    }
+};
+
+template <typename HashMap>
+struct PartitionHashMapWithOneNullableStringKey : public PartitionHashMapBase {
+    using Iterator = typename HashMap::iterator;
+    HashMap hash_map;
+    PartitionChunks null_key_value;
+
+    PartitionHashMapWithOneNullableStringKey(int32_t chunk_size) : PartitionHashMapBase(chunk_size) {}
+
+    void append_chunk(ChunkPtr chunk, const Columns& key_columns) {
+        DCHECK(key_columns[0]->is_nullable());
+        const auto* nullable_key_column = ColumnHelper::as_raw_column<NullableColumn>(key_columns[0].get());
+        const auto* key_column = down_cast<BinaryColumn*>(nullable_key_column->data_column().get());
+        append_chunk_for_one_nullable_key(hash_map, null_key_value, chunk, nullable_key_column,
+                                          [&](uint32_t offset) { return key_column->get_slice(offset); });
+    }
+};
+
+// TODO(hcf) to be implemented
+template <typename HashMap>
```

Comment:
This has not been implemented yet. If we have a multi-column Partition, is it avoided on FE now?

Reply:
Maybe multi-column partition don't need this optimize, the performance of multi column hash isn't good @liuyehcf

Reply:
Why not? If multiple columns are not supported, the benefit will be limited.

Reply:
I plan to implement multi-column in the next pr because it's priority is not high.